### PR TITLE
MVP for using rust-lld as part of cc

### LINKED ISSUE #71519: Support feeding rust-lld into gcc (and clang)
This is a subtask of #71515 

## summary of discussion

On linux/unix platforms, you aren't supposed to directly invoke ld/lld. You're supposed to invoke the linker through your system c compiler (i.e. gcc), whose responsibility it is to discover system symbols like crt1.o and provide them to ld. This means we can't "just" use rust-lld; we must feed it into gcc/clang/whatever. (We *really* don't want to implement this system symbol logic ourselves.)

In general, you can't provide the linker as a path, you must inject it into the C compiler's search path as "ld". ("ld.lld" with `-fuse-ld=lld` would also be possible, but requires newer versions of GCC, so plain "ld" is preferable if possible.)

Solution: add a new [-C linker-flavor](https://doc.rust-lang.org/rustc/codegen-options/index.html#linker-flavor) mode, `gcc-lld`, that would place rust-lld on the PATH as "ld" and invoke it through the system cc.

Note that distros do not build rust's private rust-lld binary. They make their rust packages depend on lld and symlink it in. So if you have installed rustc via your package manager, using rust-lld *is* using your system lld. This is all to say that it's generally unnecessary to "check" for a system lld, as rust-lld will already be pointing at it for the folks who care about using system toolchains. 

Should this mode become the default on linux, anyone slipping through the cracks can similarly replace rust-lld with a symlink, or explicitly pass a linker-flavor.

Note: the private rust-lld binary can be found at: 

`$(rustc --print sysroot)/lib/rustlib/$HOST_TARGET_TRIPLE/bin/rust-lld` 

which for desktop linux would be:

`$(rustc --print sysroot)/lib/rustlib/x86_64-unknown-linux-gnu/rust-lld`

## original comment:

On linux/unix platforms, you aren't supposed to directly invoke ld/lld. You're supposed to invoke the linker through your system c compiler (i.e. gcc), whose responsibility it is to discover system symbols like crt1.o and provide them to ld. This means we can't "just" use rust-lld; we must feed it into gcc/clang/whatever. (We *really* don't want to implement this system symbol logic ourselves.)

In general, you can't provide the linker as a path, you must inject it into the C compiler's search path as "ld". Alternatively, you can do the same thing but inject it as "ld.lld", and pass "-fuse-ld=lld" to gcc, which may be important as apparently lld does clang-style binary name detection to select different behaviour based on whether it's "ld" or "ld.lld". **(needs investigation)**

Unfortunately `-fuse-ld=lld` is only part of GCC 9, so we may require feature/version detection to use it (clang has had it for a long time). Doing this detection may be fairly expensive (execing gcc, possibly multiple times), so we may not want to do this in rustc itself. Ideally something higher up like cargo or maybe even rustup or something would do this analysis and cache the result. **(needs design discussion)**

It's also not clear to me how one would manually opt into this compilation mode, and/or force it on. Should it be controlled by `-C linker-flavor`? **(needs design discussion)**

### PR DESCRIPTION
**Pull request by sledgehammervampire**
Will fix #71519. I need to figure out how to write a test showing that lld is used instead of whatever linker cc normally uses. When I manually run rustc using `echo 'fn main() {}' | RUSTC_LOG=rustc_codegen_ssa::back::link=debug ./rustc -Clinker-flavor=gcc-lld --crate-type bin -Clink-arg=-Wl,-v` (thanks to bjorn3 on Zulip), I can see that lld is used, but I'm not sure how to inspect that output in a test.

### CODE REVIEW

**Comment by 1000teslas** [comment]
Hmm, it looks like the test passed because I had `lld = true` in my config.toml and used the rust-lld built by `x.py`.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 1]
```diff
@@ -1614,6 +1615,26 @@ fn linker_with_args<'a, B: ArchiveBuilder<'a>>(
...
 
+    if let LinkerFlavor::Lld(LldFlavor::Gcc) = flavor {
+        #[cfg(target_os = "linux")]
+        {
+            let tools_path =
+                sess.host_filesearch(PathKind::All).get_tools_search_paths(crt_objects_fallback);
+            // FIXME handle rust-lld not found
+            let lld_path = tools_path.into_iter().find(|p| p.join("rust-lld").exists()).unwrap();
+            // FIXME handle error instead of silently ignoring
+            let _ = std::os::unix::fs::symlink(lld_path.join("rust-lld"), lld_path.join("ld"));
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/link.rs:1626]: You can't just create a symlink in the sysroot here. It may be read-only if rustc is installed globally. Even if it isn't, you shouldn't silently modify the sysroot. Instead you should probably create this symlink in the build system during the dist step for rust-lld.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 2]
```diff
@@ -1614,6 +1615,26 @@ fn linker_with_args<'a, B: ArchiveBuilder<'a>>(
     let cmd = &mut *codegen_results.linker_info.to_linker(base_cmd, &sess, flavor);
     let link_output_kind = link_output_kind(sess, crate_type);
 
+    if let LinkerFlavor::Lld(LldFlavor::Gcc) = flavor {
+        #[cfg(target_os = "linux")]
+        {
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/link.rs:1620]: I don't think it is necessary to conditionally compile this code for linux.

File: compiler/rustc_target/src/spec/mod.rs [hunk 3]
```diff
@@ -122,6 +123,7 @@ impl ToJson for LldFlavor {
             LldFlavor::Ld => "gnu",
             LldFlavor::Link => "link",
             LldFlavor::Wasm => "wasm",
+            LldFlavor::Gcc => todo!(),
```
Comment (bjorn3) [inline compiler/rustc_target/src/spec/mod.rs:126]:
```suggestion
            LldFlavor::Gcc => "gcc",
```
and also add it to the `from_str` impl above.

File: compiler/rustc_target/src/spec/mod.rs [hunk 4]
```diff
@@ -165,6 +167,7 @@ flavor_mappings! {
     ((LinkerFlavor::Lld(LldFlavor::Ld64)), "ld64.lld"),
     ((LinkerFlavor::Lld(LldFlavor::Ld)), "ld.lld"),
     ((LinkerFlavor::Lld(LldFlavor::Link)), "lld-link"),
+    ((LinkerFlavor::Lld(LldFlavor::Gcc)), "gcc-lld"),
```
Comment (bjorn3) [inline compiler/rustc_target/src/spec/mod.rs:170]: Maybe this should be `lld-gcc`.

Comment (1000teslas) [reply compiler/rustc_target/src/spec/mod.rs:170]: I'm using the name given by @Gankra, though I don't feel particularly strongly about it.

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 1 above)
Comment (1000teslas) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1626]: @bjorn3 Sorry, I'm not sure what you mean by build system. Do you mean the directory my clone of the repo is in? Also, by dist step, do you mean in `src/bootstrap/dist.rs`?

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1626]: > I'm not sure what you mean by build system.

The build system is the code in `src/bootstrap`.

> Also, by dist step, do you mean in src/bootstrap/dist.rs?

Yes

Comment (1000teslas) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1626]: Wait, how would that work? Suppose I'm using a copy of rustc installed as a binary from my distro. Where should I create the symlink then?

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1626]: The symlink should be in `$SYSROOT/lib/rustlib/$TARGET/bin/` next to `rust-lld`.

Comment (1000teslas) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1626]: Huh, I remember using your suggestion to find the sysroot [here](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/How.20to.20get.20sysroot.20and.20HOST_TARGET_TRIPLE), but then I discovered that snippet produced identical results on my machine without having to refactor any code. When would the two differ?

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1626]: I don't think they would. I was just using a more explicit path in my comment. The point is that the ld symlink should be in the same dir as rust-lld.

Comment (1000teslas) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1626]: Wait, so what would I need to do then if your suggestion was the same as my original path?

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1626]: The suggestion is to create the symlink in src/bootstrap/dist.rs at build time instead of in link.rs at runtime.

Comment (1000teslas) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1626]: I see `dist.rs` prepares a tarball of rust for final packaging. But when I am testing, if I create the symlink in `dist.rs`, will the symlink be created for the test?

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/back/link.rs]: No, it won't. My bad. It should be in `src/bootstrap/compile.rs` after the copy just after `if let Some(lld_install) = lld_install {` I think.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 5]
```diff
@@ -1614,6 +1615,19 @@ fn linker_with_args<'a, B: ArchiveBuilder<'a>>(
...
+            sess.host_filesearch(PathKind::All).get_tools_search_paths(crt_objects_fallback);
+        // FIXME handle rust-lld not found
+        let lld_path = tools_path.into_iter().find(|p| p.join("rust-lld").exists()).unwrap();
+        cmd.arg({
+            let mut arg = OsString::from("-B");
+            arg.push(lld_path);
+            arg
+        });
+        debug!("{:?}", cmd.cmd().get_args());
+    }
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/link.rs:1629]: Should this be part of `linker_info.to_linker()`?

File: src/bootstrap/compile.rs [hunk 6]
```diff
@@ -1102,6 +1102,9 @@ impl Step for Assemble {
             let src_exe = exe("lld", target_compiler.host);
             let dst_exe = exe("rust-lld", target_compiler.host);
             builder.copy(&lld_install.join("bin").join(&src_exe), &libdir_bin.join(&dst_exe));
+            // FIXME handle error instead of silently ignoring
+            #[cfg(target_os = "linux")]
+            let _ = std::os::unix::fs::symlink(libdir_bin.join(&dst_exe), libdir_bin.join("ld"));
```
Comment (bjorn3) [inline src/bootstrap/compile.rs:1107]: The target os may not be the same as the build os, so `#[cfg(target_os = "linux")]` is wrong. I think the `#[cfg(target_os = "linux")]` should even be completely removed. It would likely give problems on windows though due to symlinks only being allowed for non-privileged users when on windows 10 with the developer mode enabled.

Comment (1000teslas) [reply src/bootstrap/compile.rs:1107]: oh right, target_os is what we build the compiler for, but we do this step at build time now. what's the preferred way of making symlinks?

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 5 above)
Comment (1000teslas) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1629]: sorry, i don't understand your question. what should be part of `linker_info.to_linker()`? the debug output?

File: src/bootstrap/compile.rs (hunk 6 above)
Comment (bjorn3) [reply src/bootstrap/compile.rs:1107]: I am not sure. cc @Mark-Simulacrum?

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 5 above)
Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1629]: This whole block to add the `-B` arg to the linker command.

File: src/bootstrap/compile.rs (hunk 6 above)
Comment (1000teslas) [reply src/bootstrap/compile.rs:1107]: wait no, the `target_os` is for the build machine of the compiler right? according to the replacements of `https://doc.rust-lang.org/std/fs/fn.soft_link.html`, maybe a `#[cfg(unix)]` with `std::os::unix::fs::symlink` and a `#[cfg(windows)]` with `std::os::windows::fs::symlink_file`?

Comment (bjorn3) [reply src/bootstrap/compile.rs:1107]: `symlink_file` requires admin permission when developer mode is not enabled or when not running on windows 10.

Comment (bjorn3) [reply src/bootstrap/compile.rs:1107]: Maybe `ld.lld` could be a small rust program running `lld` with `-flavor gnu` as first two args?

Comment (1000teslas) [reply src/bootstrap/compile.rs]: what about hard links?

Comment (bjorn3) [reply src/bootstrap/compile.rs]: That could work I think, though I am not sure if tar preserves hard links. You can try. The `builder.copy()` function tries to hardlink first by the way before falling back to copying.

Comment (1000teslas) [reply src/bootstrap/compile.rs]: looks like hard links are preserved https://unix.stackexchange.com/questions/43037/dereferencing-hard-links. actually, yeah, copying sounds like a good idea.

File: src/bootstrap/compile.rs [hunk 7]
```diff
@@ -1102,6 +1102,10 @@ impl Step for Assemble {
             let src_exe = exe("lld", target_compiler.host);
             let dst_exe = exe("rust-lld", target_compiler.host);
             builder.copy(&lld_install.join("bin").join(&src_exe), &libdir_bin.join(&dst_exe));
+            // we only support gcc-lld for x86_64-linux right now
+            if target_compiler.host.triple == "x86_64-unknown-linux-gnu" {
+                builder.copy(&libdir_bin.join(&dst_exe), &libdir_bin.join("ld"));
+            }
```
Comment (bjorn3) [inline src/bootstrap/compile.rs:1108]: I think it should be fine to just copy it for all host targets.

Comment (1000teslas) [reply src/bootstrap/compile.rs:1108]: does gcc run on windows? if so, does it look for ld.exe instead of ld?

Comment (bjorn3) [reply src/bootstrap/compile.rs:1108]: Yeah, it can run on windows. It probably indeed looks for ld.exe then.

File: src/bootstrap/compile.rs [hunk 8]
```diff
@@ -1102,6 +1102,11 @@ impl Step for Assemble {
             let src_exe = exe("lld", target_compiler.host);
             let dst_exe = exe("rust-lld", target_compiler.host);
             builder.copy(&lld_install.join("bin").join(&src_exe), &libdir_bin.join(&dst_exe));
+            // for gcc-lld linker flavor
+            builder.copy(
+                &libdir_bin.join(&dst_exe),
```
Comment (bjorn3) [inline src/bootstrap/compile.rs:1107]:
```suggestion
                &lld_install.join("bin").join(&src_exe),
```
This is a bit nicer IMHO.

File: compiler/rustc_codegen_ssa/src/back/linker.rs [hunk 9]
```diff
@@ -70,6 +74,23 @@ impl LinkerInfo {
                     as Box<dyn Linker>
             }
 
+            LinkerFlavor::Lld(LldFlavor::Gcc) => {
+                let mut cmd = cmd;
+                let tools_path =
+                    sess.host_filesearch(PathKind::All).get_tools_search_paths(self_contained);
+                // FIXME handle rust-lld not found
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/linker.rs:81]: You can use `sess.err()` when `ld` doesn't exist.

File: compiler/rustc_codegen_ssa/src/back/linker.rs [hunk 10]
```diff
@@ -70,6 +74,23 @@ impl LinkerInfo {
...
+                let lld_path =
+                    tools_path.into_iter().find(|p| p.join("rust-lld").exists()).unwrap();
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/linker.rs:83]: Please search for the actual binary that will run (`ld`)

File: src/test/ui/linkage-attr/issue-71519.rs [hunk 11]
```diff
@@ -0,0 +1,9 @@
+// test linking using cc with rust-lld injected into search path as ld
+// see rust-lang/rust#71519 for more info
+
+// build-pass
+// only-x86_64-unknown-linux-gnu
+// compile-flags: -C linker-flavor=gcc-lld
+// needs-rust-lld
```
Comment (bjorn3) [inline src/test/ui/linkage-attr/issue-71519.rs:7]: I don't think this test will fail when lld is not found.

**Comment by bjorn3** [comment]
Should be fine once these three review comments are fixed I think.

File: src/test/ui/linkage-attr/issue-71519.rs (hunk 11 above)
Comment (1000teslas) [reply src/test/ui/linkage-attr/issue-71519.rs:7]: I ignored the test in `src/tools/compiletest/src/header.rs` if rust-lld couldn't be found. I'm still not really satisfied with the test, however, since the test could link using another linker and still pass. I'm not sure how I could check the output as you mentioned [here](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/How.20to.20check.20linker.20used.20by.20cc) in a test. Also, I wonder if my test is too trivial. Would there be any weird interactions with larger examples involving multiple crates, etc.?

Comment (bjorn3) [reply src/test/ui/linkage-attr/issue-71519.rs:7]: I mean if rust-lld exists and thus the test gets run, but gcc doesn't find/use it for some reason this test will still pass.

Comment (1000teslas) [reply src/test/ui/linkage-attr/issue-71519.rs:7]: How would you suggest testing it? It seems like this is not something someone would usually encounter, since the output of linkers are designed to be interchangeable. Also, I couldn't figure out how to extract the output from `-Clink-arg=-Wl,-v` when I entered it in compile-flags, since it seems like the output only gets captured if the test fails.

Comment (bjorn3) [reply src/test/ui/linkage-attr/issue-71519.rs:7]: You could add a `run-make` test. That gives you the ability to run arbitrary commands, including checking the rustc output. You could write the output of rustc to a file and then use `$(CGREP)` to check if the output contains a given string. See for example `src/test/run-make/env-dep-info/Makefile` on how to use `$(CGREP)`.

**Comment by petrochenkov** [comment]
Using `-C linker-flavor` for this seems hacky, at least internally.
`-C linker-flavor` is about command line interface provided by the linker, but "gcc-lld" provides exactly the same interface as "gcc".
(In this sense separating e.g. "ld" and "ld.lld" flavors is already a problem, you have to always ensure that they behave identically when used in the compiler code.)

Additionally, for all targets that currently add options to `LinkerFlavor::Gcc` linkers those options need to be duplicated with the `LinkerFlavor::LldFlavor(LldFlavor::Gcc)` linker flavor as well, otherwise they won't work correctly. (That's a lot of duplication.)

If you want to keep the `-C linker-flavor` for the surface command line syntax, instead of introducing a new option, then one alternative would be to introduce a new internal enum for linker flavor used for target specs and other use cases related to linker interface. The surface linker flavor enum would then only be responsible for the stuff in `LinkerInfo::to_linker`/`linker_and_flavor`.

File: src/test/run-make/issue-71519/Makefile [hunk 12]
```diff
@@ -0,0 +1,6 @@
+-include ../../run-make-fulldeps/tools.mk
+
+# needs-rust-lld
+all:
+	RUSTC_LOG=rustc_codegen_ssa::back::link=info $(RUSTC) -C linker-flavor=gcc-lld -C link-args=-Wl,-v main.rs 2> $(TMPDIR)/output.txt
+	$(CGREP) -e "^LLD [0-9]+\.[0-9]+\.[0-9]+ \(https://github.com/rust-lang/llvm-project.git" < $(TMPDIR)/output.txt
```
Comment (bjorn3) [inline src/test/run-make/issue-71519/Makefile:6]: nit: missing trailing newline

**Comment by 1000teslas** [comment]
@petrochenkov I am new to the "guts" of Rust, so I must admit I am rather stumped by your comment.

> `-C linker-flavor` is about command line interface provided by the linker, but "gcc-lld" provides exactly the same interface as "gcc".
What do you mean by interface? Do you mean the arguments passed to the linker?

> (In this sense separating e.g. "ld" and "ld.lld" flavors is already a problem, you have to always ensure that they behave identically when used in the compiler code.)
Could you elaborate on this?

> those options need to be duplicated
Yes, indeed, that seems like a pain.

**Comment by petrochenkov** [comment]
@1000teslas 
>What do you mean by interface? Do you mean the arguments passed to the linker?

Yes, which kind of command line syntax that linker accepts.

>Could you elaborate on this?

I mean `ld` and `ld.lld` are interface-compatible, so `rustc` is supposed to always treat them identically, with exception of retrieving the linker executable "name" (`ld` vs `lld -flavor=ld`). Same is true for `gcc` and `gcc-lld`.

>Yes, indeed, that seems like a pain. How about something like LinkerFlavor::Gcc(Ld) where enum Ld { Default, Lld }?

Could you use a separate new `-Z` option for the initial version instead of a new linker flavor?
It will keep all target specs working and will help to focus the PR on more substantial issues in general.
Also a `-Z` flag will automatically provide feature gating for this new functionality.
User-facing interface can be discussed later.

**Review by bjorn3 — APPROVED** [review]
The CI changes will need to be reverted before merge, but otherwise LGTM.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 13]
```diff
@@ -1766,6 +1766,8 @@ fn linker_with_args<'a, B: ArchiveBuilder<'a>>(
     // FIXME: In practice built-in target specs use this for arbitrary order-independent options,
     // introduce a target spec option for order-independent linker options and migrate built-in
     // specs to it.
+
+    // NO-OPT-OUT, OBJECT-FILES-MAYBE, CUSTOMIZATION-POINT
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:1770]: Nit: unnecessary change.

File: src/ci/docker/host-x86_64/x86_64-gnu/Dockerfile [hunk 14]
```diff
@@ -25,4 +25,5 @@ ENV RUST_CONFIGURE_ARGS \
  --enable-sanitizers \
  --enable-profiler \
  --enable-compiler-docs
+
```
Comment (petrochenkov) [inline src/ci/docker/host-x86_64/x86_64-gnu/Dockerfile:28]: Nit: unnecessary change.

File: src/bootstrap/compile.rs [hunk 15]
```diff
@@ -1108,6 +1108,11 @@ impl Step for Assemble {
             let src_exe = exe("lld", target_compiler.host);
             let dst_exe = exe("rust-lld", target_compiler.host);
             builder.copy(&lld_install.join("bin").join(&src_exe), &libdir_bin.join(&dst_exe));
+            // for gcc-lld linker flavor
+            builder.copy(
+                &lld_install.join("bin").join(&src_exe),
+                &libdir_bin.join(exe("ld", target_compiler.host)),
+            );
```
Comment (petrochenkov) [inline src/bootstrap/compile.rs:1115]: `src\bootstrap\dist.rs` has analogous logic of copying `rust-lld` to `lib/rustlib/<host>/bin`, this bit should probably added there as well.

File: src/bootstrap/compile.rs [hunk 16]
```diff
@@ -1108,6 +1108,11 @@ impl Step for Assemble {
             let src_exe = exe("lld", target_compiler.host);
             let dst_exe = exe("rust-lld", target_compiler.host);
             builder.copy(&lld_install.join("bin").join(&src_exe), &libdir_bin.join(&dst_exe));
+            // for gcc-lld linker flavor
+            builder.copy(
+                &lld_install.join("bin").join(&src_exe),
+                &libdir_bin.join(exe("ld", target_compiler.host)),
```
Comment (petrochenkov) [inline src/bootstrap/compile.rs:1114]: The comment above says:
```rust
// We prepend this bin directory to the user PATH when linking Rust binaries. To
// avoid shadowing the system LLD we rename the LLD we provide to `rust-lld`.
```
Namely, the `lib/rustlib/<host>/bin` directory is added to `PATH` in `fn get_linker`.

This means `ld` (or `ld.exe`) will interfere with gcc's linker search even if `-Zgcc-ld=lld` is not passed.

GCC prefers hardcoded paths like `/mingw64/x86_64-w64-mingw32/bin` (on windows-gnu) to searching `PATH` so the right ld was chosen in my case, but this newly added `ld` can still interfere with search in other cases.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 17]
```diff
@@ -2495,3 +2499,36 @@ fn get_apple_sdk_root(sdk_name: &str) -> Result<String, String> {
...
+                    cmd.cmd().arg({
+                        let mut arg = OsString::from("-B");
+                        arg.push(lld_path);
+                        arg
+                    });
+                    debug!("{:?}", cmd.cmd().get_args());
+                }
+            }
+        } else {
+            sess.fatal("option gcc_ld is selected even though linker flavor is not gcc");
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:2531]:
```suggestion
            sess.fatal("option `-Z gcc-ld` is used even though linker flavor is not gcc");
```

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 18]
```diff
@@ -2495,3 +2499,36 @@ fn get_apple_sdk_root(sdk_name: &str) -> Result<String, String> {
...
+                        .find(|p| {
+                            p.join(if sess.host.is_like_windows { "ld.exe" } else { "ld" }).exists()
+                        })
+                        .unwrap_or_else(|| sess.fatal("rust-lld (as ld) not found"));
+                    cmd.cmd().arg({
+                        let mut arg = OsString::from("-B");
+                        arg.push(lld_path);
+                        arg
+                    });
+                    debug!("{:?}", cmd.cmd().get_args());
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:2527]: Nit: some debug logging is left.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 19]
```diff
@@ -2495,3 +2499,36 @@ fn get_apple_sdk_root(sdk_name: &str) -> Result<String, String> {
...
+    crt_objects_fallback: bool,
+    flavor: LinkerFlavor,
+) {
+    if let Some(ld_impl) = sess.opts.debugging_opts.gcc_ld {
+        if let LinkerFlavor::Gcc = flavor {
+            match ld_impl {
+                LdImpl::Lld => {
+                    let tools_path = sess
+                        .host_filesearch(PathKind::All)
+                        .get_tools_search_paths(crt_objects_fallback);
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:2515]:
```suggestion
                        .get_tools_search_paths(false);
```
The newly added linker file is never placed to the `self-contained` directory.

**Comment by petrochenkov** [comment]
I've tested this on windows-gnu and it looks like it works.
There are some issues with the newly added `ld` always being in `PATH` though.

File: src/bootstrap/compile.rs (hunk 16 above)
Comment (petrochenkov) [reply src/bootstrap/compile.rs:1114]: I think this can be addressed by placing `ld`/`ld.exe` to a new subdirectory in `lib/rustlib/<host>/bin` (similarly to `lib/rustlib/<host>/bin/self-contained`), which is never added to `PATH` and is only used in `-Bprefix` when `-Z gcc-ld=lld` is enabled.

File: src/bootstrap/dist.rs [hunk 20]
```diff
@@ -402,6 +402,10 @@ impl Step for Rustc {
             if builder.config.lld_enabled {
                 let exe = exe("rust-lld", compiler.host);
                 builder.copy(&src_dir.join(&exe), &dst_dir.join(&exe));
+                // for `-Z gcc-ld=lld`
+                let gcc_lld_dir = dst_dir.join("gcc_ld");
```
Comment (petrochenkov) [inline src/bootstrap/dist.rs:406]:
```suggestion
                let gcc_lld_dir = dst_dir.join("gcc-ld");
```
Nit: for style compatibility with naming in other places.

**Comment by petrochenkov** [comment]
r=me with the directory renamed (https://github.com/rust-lang/rust/pull/85961#discussion_r648129904) and commits squashed.

r? @petrochenkov

**Comment by 1000teslas** [comment]
I've fixed the conflict (it was just merging the import sets from the new master and my PR).

**Comment by 1000teslas** [comment]
@petrochenkov I was trying to dogfood my own PR, and I realized that I copied "rust-lld" as "rust-lld" instead of "ld" in gcc_ld_dir in `src/bootstrap/dist.rs`. Should I make a new PR?

**Comment by petrochenkov** [comment]
@1000teslas 
Yes, pleas make a new PR.

>I get errors that look like

Looks like the file passes the `exists` check, but then cannot actually be executed.
Could you check whether it has executable rights?

**Comment by 1000teslas** [comment]
@petrochenkov

> Looks like the file passes the exists check, but then cannot actually be executed. Could you check whether it has executable rights?

Are you talking about the copy of `rust-lld` as `ld`? If so, then yes.
